### ISSUE TITLE
[5.4.23]  Resource controller methods do not exist

### ISSUE THREAD

**Issue by WingmanImd** [laravel/framework#19366]
- Laravel Version: 5.4.23
- PHP Version: 7.1
- Database Driver & Version:

### Description:
Resource controllers do not work after update to 5.4.23

Error: **ReflectionException in RouteSignatureParameters.php line 39:
Method App\Http\Controllers\AdController::create() does not exist**

None of the controller methods are working.
It worked before using **5.4.15**

If it mathers I'm using homestead.

### Steps To Reproduce:

web.php
`Route::resource('ads', 'AdController');`  - it's the only route I have

AdController.php (yes it's in the right path, I checked 10 times)
```
<?php
namespace App\Http\Controllers;
use App\Ad;
use Illuminate\Http\Request;

class AdController extends Controller {
	/**
	 * Display a listing of the resource.
	 *
	 * @return \Illuminate\Http\Response
	 */
	public function index() {
		return Ad::all();
	}

	/**
	 * Show the form for creating a new resource.
	 *
	 * @return \Illuminate\Http\Response
	 */
	public function create() {
		return view('frontend.pages.new_ad');
	}

	/**
	 * Store a newly created resource in storage.
	 *
	 * @param  \Illuminate\Http\Request  $request
	 * @return \Illuminate\Http\Response
	 */
	public function store(Request $request) {
		dump('redirect to ad page');
		return redirect()->back();
	}

	/**
	 * Display the specified resource.
	 *
	 * @param  \App\Ad  $ad
	 * @return \Illuminate\Http\Response
	 */
	public function show(Ad $ad) {
		return view('frontend.pages.ad')->with(['ad' => $ad]);
	}

	/**
	 * Show the form for editing the specified resource.
	 *
	 * @param  \App\Ad  $ad
	 * @return \Illuminate\Http\Response
	 */
	public function edit(Ad $ad) {
		//
	}

	/**
	 * Update the specified resource in storage.
	 *
	 * @param  \Illuminate\Http\Request  $request
	 * @param  \App\Ad  $ad
	 * @return \Illuminate\Http\Response
	 */
	public function update(Request $request, Ad $ad) {
		//
	}

	/**
	 * Remove the specified resource from storage.
	 *
	 * @param  \App\Ad  $ad
	 * @return \Illuminate\Http\Response
	 */
	public function destroy(Ad $ad) {
		//
	}
}
```

Result of the composer update (maybe it helps, because it's a really weird bug)

  ```
- Updating symfony/css-selector (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/var-dumper (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/routing (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/process (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/http-foundation (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/event-dispatcher (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/debug (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/http-kernel (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/finder (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating symfony/console (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating swiftmailer/swiftmailer (v5.4.6 => v5.4.8): Downloading (100%)
  - Updating ramsey/uuid (3.5.2 => 3.6.1): Downloading (100%)
  - Updating symfony/translation (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating league/flysystem (1.0.35 => 1.0.40): Downloading (100%)
  - Updating erusev/parsedown (1.6.1 => 1.6.2): Downloading (100%)
  - Updating laravel/framework (v5.4.15 => v5.4.23): Downloading (100%)
  - Updating laravelcollective/html (v5.4.1 => v5.4.8): Downloading (100%)
  - Updating symfony/yaml (v3.2.6 => v3.2.8): Downloading (100%)
  - Updating sebastian/diff (1.4.1 => 1.4.3): Downloading (100%)
  - Updating phpunit/php-code-coverage (4.0.7 => 4.0.8): Downloading (100%)
  - Updating myclabs/deep-copy (1.6.0 => 1.6.1): Downloading (100%)
  - Updating phpunit/phpunit (5.7.15 => 5.7.20): Downloading (100%)
  - Updating guzzlehttp/psr7 (1.4.1 => 1.4.2): Downloading (100%)
  - Updating psy/psysh (v0.8.2 => v0.8.5): Downloading (100%)
```

### How I "fixed" it:

I created a new controller with artisan called "TestController" and edited to route to user TestController instead of AdController. (Sound stupid heh? yet it worked )

**Comment by sisve** [comment]
I am unable to reproduce this on 5.4.23. You have a syntax error on your namespace declaration, but did produced another error.

    namespace App\Http\Controllers;`

Can you provide a test that can be used to reproduce the problem?

**Comment by WingmanImd** [comment]
@sisve The syntax error is from writing the issue on github sorry.
The exact(or as close as I can remeber) stept I made are:
- Start homestead
- php artisan make:controller Admin\BlogController --resource --model=BlogPost
- This created an AdminBlogController.php file in the App\Http\Controllers
- Renamed the file to BlogController .php and moved inside App\Http\Controllers\Admin\
- namespace error(duh)
- Moved it back and renamed it AdminBlogController.php
- Created an resource route on it and none of the model bindings didn't work(another weird bug)
Ex: `public function edit(Ad $ad) {}` $ad would return and empty Ad model and not the model with that id....
- composer update, still the model binding didn't work
- frontend stuff
- Visited the Ad page and it didn't work anymore, even though I didn't edit anything about it.

I just said all I can remember, maybe you know something I don't know. :D

**Comment by WingmanImd** [comment]
An update here: Creating another controller(with another name) and mapping the route to it works but if I rename the new Controller to "AdController.php" causes the same error. This is really weird.

**Comment by WingmanImd** [comment]
Ok, found out I made a mistake, but it got even weirder.
I had an AdminAdController.php with `class AdController extends Controller {` 
I fixed it and now I get:
```
ReflectionException in Container.php line 721:
Class App\Http\Controllers\AdController does not exist
```
The file exists and it's defined properly.

So I guess the way to reproduce this is:

1. Duplicate Controller.php and change it's name to AdController(also in it's class name)
2. Rename the file to AdAdminController but don't change it's class name
3 `php artisan make:controller AdController --resource --model=Ad`
4. add `Route::resource('ads', 'AdController');` in web.php

**Comment by WingmanImd** [comment]
Seems like it was an autoload problem.

I fixed it by running `composer dump-autoload -o`

**Comment by Kyslik** [comment]
If that is the case you should **double** check your namespaces since using -o flag means that composer does some magic (looking inside classes etc.) in order to generate valid autoloader file.

**Comment by WingmanImd** [comment]
@Kyslik What would you suggest to use instead?

**Comment by Kyslik** [comment]
I am simply stating, if `cdu` results in invalid autoloader file then you have typo somewhere.

(composer dump-autoload as cdu)

**Comment by WingmanImd** [comment]
@Kyslik Thanks!

**Comment by pratik007kumar** [comment]
i have the same problem but not working for me
namespace App\Http\Controllers;

```
use App\Http\Controllers\User\LoginController;
use App\User;
use App\User_information;
use Illuminate\Http\Request;
use Lang;
use View;
use Session;
use Config;
use Auth;
class FrontendRegisterController extends Controller {
public function abc()
{
	echo "string";
}
}
```
`Route::get('abc', 'FrontendRegisterController@abc');
`

`
ReflectionException in RouteSignatureParameters.php line 39: Method App\Http\Controllers\FrontendRegisterController::abc() does not exist`

**Comment by aheropcs** [comment]
Here is another example:

My co-worker used Been to edit the project but for some reason created a copy of the file, everything looks for OK until I have to start to do the cleanup and the system throughout the same error, I found the solution deleting all the copy files and run as @WingmanImd indicated lines above.

Hope this bug can be fixed next time for somebody else without expending many hours.

"Seems like it was an autoload problem.I fixed it by running composer dump-autoload -o"
--

**Comment by Kyslik** [comment]
@aheropcs If you can fix your problem by using `--optimize`, you have a typo somewhere, or you do not follow the standards.

**Comment by hondaman900** [comment]
I had this same "Method not found in controller" issue after running composer update. This was very frustrating as I could see the method was actually there, just not being recognized by Laravel.

The issue was I had copied my controller file during development in the past, renaming the copied version, but then moved on and forgot about the renamed copied version. Since then I added a new method to the original controller, which was working fine. Then when composer updated during the install of a package, it found conflicting classes named the same in both controllers and gave priority to the older one, thus rendering my new controller and method as unavailable/invisible. Running `composer-dump-autoload -o` only worked **after I deleted the copied controller**, but once I did delete it and ran `composer-dump-autoload -o`, my method reappeared and the conflict was resolved.

Hope this helps.
